### ISSUE TITLE
Research to add support to JSON Patch Schema (RFC 6902)

### ISSUE THREAD

**Issue by ricardozanini** [serverlessworkflow/specification#149]
**What would you like to be added**:
From the [RFC 6902](https://tools.ietf.org/html/rfc6902) document:

>  JSON Patch defines a JSON document structure for expressing a
   sequence of operations to apply to a JavaScript Object Notation
   (JSON) document

Would be interesting to support this RFC on scenarios where an user need to modify the content of the data within the workflow. I'm opening this issue to keep track of this and do some research around use cases this would be useful instead of using JSON path directly. It's super simple to understand and it's gaining adoption in the Kubernetes community, like in the [kustomize](https://kubectl.docs.kubernetes.io/pages/reference/kustomize.html#patchesjson6902) tool for instance.

**Why is this needed**:
This is an alternative to transform the data on a declarative way.

**Comment by tsurdilo** [comment]
@ricardozanini +1, do you have any ideas where the use of this could be beneficial? It seems nice for like inject state for example. For expressions we need to assure that there is a boolean result, so using JsonPath there makes more sense. wdyt?

**Comment by ricardozanini** [comment]
I was thinking on having a new state to implement this RFC actually. So we could do small transformations during the execution of the workflow since all the context is based on JSON.

I can receive a huge data from an external source and be interested in only one part, so I would delete it. Or I could call an external function and update the context based on the result that come from it. I believe that are many possibilities where we would have such state. But again, this is just brainstorming.

Instead of declaring a function and implementing it to do transformation, we would have this built in.

**Comment by tsurdilo** [comment]
I believe that question raised in https://github.com/serverlessworkflow/specification/issues/158 would be a very good use case for this. We should definitely move this forward I think.

**Comment by ricardozanini** [comment]
I'll think on some use cases and how to add this to our scenario.

**Comment by ricardozanini** [comment]
@tsurdilo @manuelstein after brainstorming a little and going through the [State Data Filter](https://github.com/serverlessworkflow/specification/blob/0.5.x/specification.md#state-data-filtering) excellent topic, I believe we could have a `dataInputTransform` and `dataOutputTransform` defined as `State Data Transformation`.

For example, for reasons that I can't control, I receive an event with the following payload:

```json
{
  "fruits": [ "apple", "orange", "pear" ],
  "drinks": ["capirinha", "whisky", "vodka", "wine"],
  "snacks": ["french fries", "jelly beans"],
  "vegetables": [
    {
      "veggieName": "potato",
      "veggieLike": true
    },
    {
      "veggieName": "broccoli",
      "veggieLike": false
    }
  ]
}
```

I don't care about `vegetables` or `snacks`, so I could **delete** then from my payload once I receive the event:

```
{
  "name": "FruitsAndDrinksOnly",
  "type": "inject",
  "stateDataTransform": {
    "dataInputPatch": [{"op": "remove", path: "/snacks"}, {"op": "remove", path: "/vegetables"}]
  },
  "transition": {
     "nextState": "someNextState"
  }
}
```

I understand that we can also apply a filter to have only what we care about, but in real-life scenarios, we might come up with a [real huge payload](https://docs.github.com/en/free-pro-team@latest/rest/reference/pulls#get-a-pull-request). 

Creating a filter that matches only the desired attributes might be cumbersome and really hard to read. So we would have options for users to decide which kind of transformation they would like to apply in each use case.

Now we would have:

```json
{
  "fruits": [ "apple", "orange", "pear" ],
  "drinks": ["capirinha", "whisky", "vodka", "wine"]
}
```
Now I want to call my `cocktail` service, but the problem is... it does not understand `drinks`, but `beverages`. :D

I could do:

```json
{
"actions":[  
    {  
       "functionRef": {
          "refName": "makeCocktail"
       },
       "actionDataTransform": {
          "dataInputPatch": [{"op": "move", "from": "/drinks", "path": "/beverages" }]
       }
    }
 ]
}
```

Then I can safely call my cocktail service, which I don't control the interface. :)

I can think of many more examples, but I want to just start a discussion now, so I think these ones should do.

What do you guys think?

**Comment by tsurdilo** [comment]
@ricardozanini i like the idea of allowing this. Here are some questions/ideas i have, hope they help:

1) In the first example, the equivalent JsonPath expression could be just "$.[fruits,drinks]" which seems simpler to understand than 
[{"op": "remove", path: "/snacks"}, {"op": "remove", path: "/vegetables"}]. I think for selections JsonPath is better suited, where as JsonPatch is more powerful for data manipulation.

2) I would look how we can have these two (selction (jsonpath) and manipulation (jsonpatch)) work together inside filters if possible. Let's see if that works instead of trying to separate them into "data filters" and "data  transforms".

3) Look if its possible to do selections and manipulations in a more domain specific way. I think in large processes having so many jsonpath/patch expressions will become really confusing. Maybe there is a good way for us to allow users to express their selectiopns/manipulations as for example (using your examples):

"Only Drinks", "No Snacks", "Rename Drinks to Beverages" and allow users to map those to a set of path/patch expressions. WDYT?

**Comment by ricardozanini** [comment]
@tsurdilo I really liked your idea! Indeed, if we come up with a structure that can be used to express them all in a simple, no-brainer way would be wonderful. How about transforming these into functions instead, so they can be reused? That's what they are, right? But this way we offer as a standard. We could have `openapi`, `filter` and `transform` functions kind maybe?

```json 
{  
"functions": [
  {
   "name": "OnlyDrinks",
   "operation": "$.[drinks]",
   "kind": "jsonPath"
  },
  {
   "name": "NoSnacks",
   "operation": "$.[fruits,drinks]",
   "kind": "jsonPath"
  },
 {
   "name": "RenameDrinksToBeverage",
   "operation": {"op": "move", "from": "/drinks", "path": "/beverages"},
   "kind": "jsonPatch"
 }
]
}
```

That's what you had in mind? We can keep the `dataFilters` as it is for small operations, but reusable functions could bring a lot of benefits for complex workflows.

So instead of repeating `$.[fruits,drinks]` every time in the definition, we just call the function. :)

**Comment by ricardozanini** [comment]
Added this topic to the next meeting minute, since it's an important topic that deserves wider attention (cc @manuelstein)

**Comment by tsurdilo** [comment]
@ricardozanini we already have a functions top-level array for defining service invocation info so using "functions" for this too is confusing imo. 
Looking at this more, externalizing makes little sense unless these can be reusable things....hmm yeah we need to think about this definitely more.

**Comment by ricardozanini** [comment]
@tsurdilo @manuelstein Based on our conversation today, I've found these approaches for this problem:

1. [Jolt](https://github.com/bazaarvoice/jolt)
2. [jsonpath-object-transform](https://www.npmjs.com/package/jsonpath-object-transform)
3. [JSONata](https://docs.jsonata.org/overview.html)

None of them is a standard. I believe we can continue with the RFC itself for now. I can do more researches, but if I haven't found a standard despite RFC 6902 for now, I don't think we will find any. Shouldn't be too hard to find it, right?

**Comment by tsurdilo** [comment]
Please see discussions on issue https://github.com/serverlessworkflow/specification/issues/216. It is related to this.

**Comment by ricardozanini** [comment]
@tsurdilo we might close this one once we decided which path to take (no pun intended). :D

**Comment by tsurdilo** [comment]
@ricardozanini so wdyt - can this one be closed?

**Comment by tsurdilo** [comment]
depends on https://github.com/serverlessworkflow/specification/issues/230

**Comment by ricardozanini** [comment]
Closed in favor of #216
